# Business Understanding

JCPenney is a large retail company that sells a wide range of products such as clothing, beauty products, home goods, and accessories. They get these products from top brands and usually sell these products at discounted prices. 

The aim of these analysis is to understand customer behaviour, how they interact with products, how satisfied they are, and which products or brands they prefer. In this analysis I will look at product ratings, review patterns, customer demographics, and discount levels to see what influences satisfaction, which products perform best, and whether discounts actually affect customer feedback.

The important data for this analysis, is username, customer id, customer age (we will calculate this from DOB), state, prices, product categories, average product ratings, and reviews, etc.

# Data understanding & preparation

In [1]:
# Data Inspection and understanding
import pandas as pd
# JC Penney Reviewers (JSON)
reviewers = pd.read_json("jcpenney_reviewers.json", lines=True)
display(reviewers.head())
print("reviewers shape:", reviewers.shape)
print("reviewers dtypes:\n", reviewers.dtypes)
# JC Penney Products (JSON) ***(Transposed)
products = pd.read_json("jcpenney_products.json", lines=True)
display(products.head(1).T)
print("products shape:", products.shape)
print("products dtypes:\n", products.dtypes)
# Products (CSV)
products_csv = pd.read_csv("products.csv")
display(products_csv.head())
print("products_csv shape:", products_csv.shape)
print("products_csv dtypes:\n", products_csv.dtypes)
print("products_csv missing values:\n", products_csv.isna().sum())
# Reviews (CSV)
reviews_csv = pd.read_csv("reviews.csv")
display(reviews_csv.head())
print("reviews_csv shape:", reviews_csv.shape)
print("reviews_csv dtypes:\n", reviews_csv.dtypes)
# Users (CSV)
users_csv = pd.read_csv("users.csv")
display(users_csv.head())
print("users_csv shape:", users_csv.shape)
print("users_csv dtypes:\n", users_csv.dtypes)

Username         DOB          State  \
0  bkpn1412  31.07.1983         Oregon   
1  gqjs4414  27.07.1998  Massachusetts   
2  eehe1434  08.08.1950          Idaho   
3  hkxj1334  03.08.1969        Florida   
4  jjbd1412  26.07.2001        Georgia   

                                            Reviewed  
0                 [cea76118f6a9110a893de2b7654319c0]  
1                 [fa04fe6c0dd5189f54fe600838da43d3]  
2                                                 []  
3  [f129b1803f447c2b1ce43508fb822810, 3b0c9bc0be6...  
4                                                 []

reviewers shape: (5000, 4)
reviewers dtypes:
 Username    object
DOB         object
State       object
Reviewed    object
dtype: object


0
uniq_id                                  b6c0b6bea69c722939585baeac73c13d
sku                                                          pp5006380337
name_title                    Alfred Dunner® Essential Pull On Capri Pant
description             You'll return to our Alfred Dunner pull-on cap...
list_price                                                          41.09
sale_price                                                          24.16
category                                                    alfred dunner
category_tree                                jcpenney|women|alfred dunner
average_product_rating                                              2.625
product_url             http://www.jcpenney.com/alfred-dunner-essentia...
product_image_urls      http://s7d9.scene7.com/is/image/JCPenney/DP122...
brand                                                       Alfred Dunner
total_number_reviews                                                    8
Reviews                 [{'User': 'fsdv4141', 'Review': 'You never hav...
Bought With             [898e42fe937a33e8ce5e900ca7a4d924, 8c02c262567...

products shape: (7982, 15)
products dtypes:
 uniq_id                    object
sku                        object
name_title                 object
description                object
list_price                 object
sale_price                 object
category                   object
category_tree              object
average_product_rating    float64
product_url                object
product_image_urls         object
brand                      object
total_number_reviews        int64
Reviews                    object
Bought With                object
dtype: object


Uniq_id           SKU  \
0  b6c0b6bea69c722939585baeac73c13d  pp5006380337   
1  93e5272c51d8cce02597e3ce67b7ad0a  pp5006380337   
2  013e320f2f2ec0cf5b3ff5418d688528  pp5006380337   
3  505e6633d81f2cb7400c0cfa0394c427  pp5006380337   
4  d969a8542122e1331e304b09f81a83f6  pp5006380337   

                                          Name  \
0  Alfred Dunner® Essential Pull On Capri Pant   
1  Alfred Dunner® Essential Pull On Capri Pant   
2  Alfred Dunner® Essential Pull On Capri Pant   
3  Alfred Dunner® Essential Pull On Capri Pant   
4  Alfred Dunner® Essential Pull On Capri Pant   

                                         Description  Price  Av_Score  
0  Youll return to our Alfred Dunner pull-on capr...  41.09     2.625  
1  Youll return to our Alfred Dunner pull-on capr...  41.09     3.000  
2  Youll return to our Alfred Dunner pull-on capr...  41.09     2.625  
3  Youll return to our Alfred Dunner pull-on capr...  41.09     3.500  
4  Youll return to our Alfred Dunner pull-on capr...  41.09     3.125

products_csv shape: (7982, 6)
products_csv dtypes:
 Uniq_id         object
SKU             object
Name            object
Description     object
Price          float64
Av_Score       float64
dtype: object
products_csv missing values:
 Uniq_id           0
SKU              67
Name              0
Description     543
Price          2166
Av_Score          0
dtype: int64


Uniq_id  Username  Score  \
0  b6c0b6bea69c722939585baeac73c13d  fsdv4141      2   
1  b6c0b6bea69c722939585baeac73c13d  krpz1113      1   
2  b6c0b6bea69c722939585baeac73c13d  mbmg3241      2   
3  b6c0b6bea69c722939585baeac73c13d  zeqg1222      0   
4  b6c0b6bea69c722939585baeac73c13d  nvfn3212      3   

                                              Review  
0  You never have to worry about the fit...Alfred...  
1  Good quality fabric. Perfect fit. Washed very ...  
2  I do not normally wear pants or capris that ha...  
3  I love these capris! They fit true to size and...  
4  This product is very comfortable and the fabri...

reviews_csv shape: (39063, 4)
reviews_csv dtypes:
 Uniq_id     object
Username    object
Score        int64
Review      object
dtype: object


Username         DOB          State
0  bkpn1412  31.07.1983         Oregon
1  gqjs4414  27.07.1998  Massachusetts
2  eehe1434  08.08.1950          Idaho
3  hkxj1334  03.08.1969        Florida
4  jjbd1412  26.07.2001        Georgia

users_csv shape: (5000, 3)
users_csv dtypes:
 Username    object
DOB         object
State       object
dtype: object


### Observations
For the above step, i have loaded all the five files into a pandas dataframe. I have inspected the data structure, and column types.
Below are the observations from the data inspection.

**jcpenney_reviewers.json:**  
Contains 5,000 entries and 4 columns. The 'Reviewed' column contains empty lists ([]), indicating some users did not leave a comment. The date column is stored as an object and I will convert it to a proper datetime format to be able to calculate age from it.

**jcpenney_products.json:**  
Contains 7,982 entries and 15 columns. Some columns such as 'list_price' and 'sale_price' are stored as objects and I will converted to numeric values. The 'category_tree' column has three different categories that I will need to be separate into distinct columns.

**products.csv:**  
Contains 7,982 entries and 6 columns. Columns are uploaded correctly, but some values are missing. The 'price' column is missing 2,166 values which is approximately 27% of the price data, and the 'description' column is missing 543 values which is approximately 7% of the description data. 

**reviews.csv:**  
Contains 39,063 entries and 4 columns. The columns are uploaded correctly, with the 'score' column stored as an integer.

**users.csv:**  
Contains 5,000 entries and 3 columns. The 'DOB' field should be converted to datetime to allow for calculation of age.

In [2]:
# Data cleaning and preparation
import pandas as pd
import numpy as np
# Reviewers (jcpenney_reviewers.json)
# Load the data
reviewers = pd.read_json("jcpenney_reviewers.json", lines=True)
# Count empty lists in Reviewed 
null_reviews = reviewers["Reviewed"].apply(lambda x: isinstance(x, list) and len(x) == 0).sum()
print("Number of null reviews before:", null_reviews)
# Note: The empty lists are 971, approximately 19% of the total list; instead of dropping I will replace [] with NaN
reviewers["Reviewed"] = reviewers["Reviewed"].apply(lambda x: np.nan if x == [] else x)
print("Number of null reviews after:", reviewers["Reviewed"].isna().sum())
# Convert lists to strings in Reviewed
reviewers["Reviewed"] = reviewers["Reviewed"].apply(lambda x: ",".join(x) if isinstance(x, list) else x)
# Convert DOB to datetime
reviewers.DOB = pd.to_datetime(reviewers.DOB, errors="coerce", dayfirst=True)
# Save to a csv file
reviewers.to_csv("jcpenney_reviewers_clean.csv", index=False)
reviewers_clean = pd.read_csv("jcpenney_reviewers_clean.csv")

# Products (jcpenney_products.json)
# Load the data
products = pd.read_json("jcpenney_products.json", lines=True)
# Convert list_price, sale_price and average_product_rating columns to numeric
products.list_price = pd.to_numeric(products.list_price, errors="coerce")
products.sale_price = pd.to_numeric(products.sale_price, errors="coerce")
products.average_product_rating = pd.to_numeric(products.average_product_rating, errors="coerce")
# Split category_tree into 3 parts and delete the category_tree column
if "category_tree" in products.columns:
    categories = products.category_tree.astype("string").str.split("|", n=2, expand=True)
    categories = categories.rename(columns={0: "category_1", 1: "category_2", 2: "category_3"})
    products = pd.concat([products.drop(columns=["category_tree"]), categories], axis=1)
# Extract User and Review from Reviews list, then drop Reviews column
products["User"] = products["Reviews"].apply(lambda x: x[0]["User"] if isinstance(x, list) and len(x) > 0 else None)
products["Review"] = products["Reviews"].apply(lambda x: x[0]["Review"] if isinstance(x, list) and len(x) > 0 else None)
products = products.drop(columns="Reviews")
# Check, save, reload to verify
products.to_csv("jcpenney_products_clean.csv", index=False)
products_clean = pd.read_csv("jcpenney_products_clean.csv")
print(products_clean.dtypes[["list_price","sale_price","average_product_rating"]])

# Products CSV (products.csv)
# Load the data
products_csv = pd.read_csv("products.csv")
# Ensure price and av_score are numeric
products_csv.Price = pd.to_numeric(products_csv.Price, errors="coerce")
products_csv.Av_Score = pd.to_numeric(products_csv.Av_Score, errors="coerce")
# Fill Description/SKU; Price with median
products_csv["Description"] = products_csv["Description"].fillna("No description available")
products_csv["SKU"] = products_csv["SKU"].fillna("Unknown SKU")
products_csv["Price"] = products_csv["Price"].fillna(products_csv["Price"].median())
# Check tail, save, reload
products_csv.to_csv("products_csv_clean.csv", index=True)
products_csv_clean = pd.read_csv("products_csv_clean.csv", index_col=0)

# Reviews CSV (reviews.csv)
# Load the data
reviews_csv = pd.read_csv("reviews.csv")
# Ensure score is numeric
reviews_csv.Score = pd.to_numeric(reviews_csv.Score, errors="coerce")
# Save & reload
reviews_csv.to_csv("reviews_csv_clean.csv", index=True)
reviews_csv_clean = pd.read_csv("reviews_csv_clean.csv", index_col=0)

# Users CSV (users.csv)
# Load the data
users_csv = pd.read_csv("users.csv")
# Convert DOB to datetime then to string
users_csv.DOB = pd.to_datetime(users_csv.DOB, errors="coerce", dayfirst=True)
users_csv.DOB = users_csv.DOB.dt.strftime("%Y-%m-%d")
# Save & reload
users_csv.to_csv("users_csv_clean.csv", index=True)
users_csv_clean = pd.read_csv("users_csv_clean.csv", index_col=0)

Number of null reviews before: 971
Number of null reviews after: 971
list_price                float64
sale_price                float64
average_product_rating    float64
dtype: object


### Comments 

As shown above, I have cleaned and prepared all the datasets. 

The main goal was to make sure each dataset was ready for merging and further analysis. I focused on correcting the data types, fixing missing values, and restructuring columns where needed based on the earlier observations.

I have now addressed all the issues noted during the inspection stage, such as converting price and date columns into the correct formats, replacing empty lists with NaN values, and filling in missing product details. 

I identified the key columns that will allow me to link the datasets together later in the analysis: unique_id and username
With these changes, the data is now consistent, clean, and ready for merging.

### Merging the data into one data file that i can use for my analysis

In [3]:
# Data merging
# Combine all cleaned datasets and prepare a unique, analysis-ready master file.
import pandas as pd
import numpy as np
from datetime import datetime
# Load cleaned datasets and renaming them to avoid confusion
revwrs = pd.read_csv("jcpenney_reviewers_clean.csv")
prod_json = pd.read_csv("jcpenney_products_clean.csv")
prod_csv = pd.read_csv("products_csv_clean.csv")
reviews = pd.read_csv("reviews_csv_clean.csv")
users = pd.read_csv("users_csv_clean.csv")
# Standardize key column names
# Use 'uniq_id' to connect product-related datasets
# Use 'Username' to connect reviewer/user-related datasets
prod_csv.rename(columns={"Uniq_id": "uniq_id"}, inplace=True)
reviews.rename(columns={"Uniq_id": "uniq_id"}, inplace=True)
# Confirm required columns exist
assert "uniq_id" in reviews.columns and "Username" in reviews.columns
assert "uniq_id" in prod_csv.columns
assert "uniq_id" in prod_json.columns
assert "Username" in users.columns and "Username" in revwrs.columns
# Merge datasets
# Start with the reviews file (contains both uniq_id and Username)
master = reviews.copy()
# Merge products CSV (uniq_id)
master = master.merge(prod_csv, on="uniq_id", how="left")
# Merge products JSON (uniq_id)
master = master.merge(prod_json, on="uniq_id", how="left", suffixes=("", "_json"))
# Merge users (Username)
master = master.merge(users, on="Username", how="left", suffixes=("", "_user"))
# Merge reviewers (Username)
master = master.merge(revwrs, on="Username", how="left", suffixes=("", "_reviewer"))
# Check merged structure
print("Master dataset shape:", master.shape)
# Save full master dataset
master.to_csv("jcpenney_master_dataset_full.csv", index=False)
# Create trimmed version with unique columns
master = pd.read_csv("jcpenney_master_dataset_full.csv")
# Select key columns for analysis
master_unique = master[[
    "Username", "DOB", "State", "uniq_id", "sku", "name_title", "description", "brand",
    "category", "category_1", "category_2", "category_3", "list_price", "sale_price",
    "product_url", "product_image_urls", "total_number_reviews", "Reviewed",
    "Bought With", "average_product_rating", "Score", "Review"]].copy()
# Add 'Age' column from DOB
master_unique["DOB"] = pd.to_datetime(master_unique["DOB"], errors="coerce")
today = pd.Timestamp(datetime.now().date())
# Determine whether birthdays have passed this year
birthday_passed = (
    (master_unique["DOB"].dt.month < today.month) |
    ((master_unique["DOB"].dt.month == today.month) & (master_unique["DOB"].dt.day <= today.day)))
# Compute age
master_unique["Age"] = today.year - master_unique["DOB"].dt.year - (~birthday_passed).astype(int)
master_unique["Age"] = master_unique["Age"].astype("Int64")
# Move Age column next to DOB
cols = master_unique.columns.tolist()
cols.insert(cols.index("DOB") + 1, cols.pop(cols.index("Age")))
master_unique = master_unique[cols]
# Fill missing sale_price with list_price
master_unique.fillna({'sale_price': master_unique['list_price']}, inplace=True)
# Save cleaned master dataset
master_unique.to_csv("jcpenney_master_dataset_unique.csv", index=False)
# Check final structure
print("Final master dataset shape:", master_unique.shape)

Master dataset shape: (39114, 34)
Final master dataset shape: (39114, 23)


# Comments

After cleaning each dataset, I merged them into a single master dataset to make analysis easier and more consistent.  

The 'reviews' dataset served as the main base for merging since it contained both 'uniq_id' (for products) and 'Username' (for users and reviewers).
Before merging, I standardized the key column names to ensure consistency, renaming 'Uniq_id' to 'uniq_id' where necessary.  

After merging all datasets, I saved a complete master version and then created a smaller, refined file with only the key columns needed for analysis.

I also added an **Age** column derived from 'DOB' and handled any remaining missing values (e.g., filling missing 'sale_price' values with 'list_price'). In this case, if sale_price = list_price, then the product has no discount.  

The final dataset, 'jcpenney_master_dataset_unique.csv', is now clean, consistent, and ready for core analysis.

# Core Analysis

In [5]:
# 1: Who are the company's customers? (Age & State)
import pandas as pd
import matplotlib.pyplot as plt
import os
# Create folders for outputs
os.makedirs("figs", exist_ok=True)
os.makedirs("tables", exist_ok=True)
# Load master dataset
master_unique = pd.read_csv("jcpenney_master_dataset_unique.csv")
# 1) AGE ANALYSIS
# (a) Distribution by single year of age
age_count = master_unique["Age"].value_counts().sort_values(ascending=False)
print("\nAge value_counts (first 3 by age):")
print(age_count.head(3))
# (b) Create age groups
bins   = [20, 30, 40, 50, 60, 70, 80]
labels = ["21–30", "31–40", "41–50", "51–60", "61–70", "71–80"]
master_unique["Age group"] = pd.cut(
    master_unique["Age"],
    bins=bins,
    labels=labels,
    include_lowest=True, 
    right=True)
# Move Age group right after Age
cols = master_unique.columns.tolist()
if "Age group" in cols and "Age" in cols:
    cols.insert(cols.index("Age") + 1, cols.pop(cols.index("Age group")))
    master_unique = master_unique[cols]
# (c) Summary table: counts and percentages by age group
age_group_counts = master_unique["Age group"].value_counts().sort_values(ascending=False)
age_group_summary = pd.DataFrame({
    "Count": age_group_counts,
    "Percentage": (age_group_counts / age_group_counts.sum() * 100).round(2).astype(str) + "%"})
# Save age group summary to CSV
age_group_summary.to_csv("tables/age_group_summary.csv", index=True)
print("Age group summary saved to tables/age_group_summary.csv")
# (d) Visualization: bar chart of age groups
plt.figure(figsize=(8, 5))
master_unique["Age group"].value_counts().sort_index().plot(
    kind="bar",
    color="skyblue",
    title="Customer Distribution by Age Group",
    edgecolor="black")
plt.xlabel("Age Group")
plt.ylabel("Number of Customers")
plt.savefig("figs/age_group_distribution.png", dpi=300)
plt.close()
print("Age group bar chart saved to figs/age_group_distribution.png")

# 2) STATE ANALYSIS
# (a) Counts per state
state_counts = master_unique["State"].value_counts().sort_values(ascending=False)
# Save state tables
state_counts.head(10).to_csv("tables/top10_states.csv")
state_counts.tail(10).to_csv("tables/bottom10_states.csv")
print("Top/bottom 10 state tables saved to tables/ folder")
# (b) Unique states count
num_states = master_unique["State"].nunique()
print("\nNumber of unique states:", num_states)
# (c) Top 20 states as percentages
state_percentages = (state_counts.head(20) / len(master_unique) * 100).round(2)
# Save top 20 state percentages
state_percentages.to_csv("tables/state_percentages_top20.csv")
print("State percentage table saved to tables/state_percentages_top20.csv")
# (d) Visualization: bar chart of top 10 states
plt.figure(figsize=(12, 6))
state_counts.head(10).plot(
    kind="bar",
    color="skyblue",
    title="Top 10 States by Number of Customers",
    edgecolor="black")
plt.xlabel("State")
plt.ylabel("Number of Customers")
plt.savefig("figs/top10_states_customers.png", dpi=300)
plt.close()
print("Top 10 states bar chart saved to figs/top10_states_customers.png")


Age value_counts (first 3 by age):
Age
62    901
26    853
72    851
Name: count, dtype: int64
Age group summary saved to tables/age_group_summary.csv
Age group bar chart saved to figs/age_group_distribution.png
Top/bottom 10 state tables saved to tables/ folder

Number of unique states: 57
State percentage table saved to tables/state_percentages_top20.csv
Top 10 states bar chart saved to figs/top10_states_customers.png


Observations from the above analysis
The age with the highest number of customers for JCPenny are ages 62, 26, and 72 with 901, 953 and 851 customers respectively. 
The age group with the highest number of customers for JCPenny is 60-70 with 20.05% and 51-60 with 19.22%. The age group with the least customers is 71-80 with 9.78% and the age group 21-30 with 13.3%.
The state with the highest customers for JCPenny are Massachussets and Kentucky with 850 and 845 customers each respectively which contributes to approximately 2.17% and 2.16% of the total customers. The states with the least customers for JCPenny are North Carolina with 499 customers and Illinois with 541 customers. There are 57 unique states with JCPenny customers, 50 from the united states and 7 from the united states territories/non-states.

In [7]:
#What products perform best? i.e. which brands/categories/products are most successful

#1.Brand
#Categorize the brands according to the rate bands and plot a graph showing the distribution 
brand_ratings = master_unique.groupby("brand")["average_product_rating"].mean().sort_values(ascending=False)
brand_ratings.to_csv("tables/all_brands.csv")
print("Saved: tables/all_brands.csv")
#Define bins and labels
bins = [0, 1, 2, 3, 4, 5]
labels = ["0–1", "1.1–2", "2.1–3", "3.1–4", "4.1–5"]
brand_bins = pd.cut(brand_ratings, bins=bins, labels=labels, include_lowest=True)
brand_counts = brand_bins.value_counts().sort_index()
brand_counts.to_csv("tables/brand_counts_by_rating_range.csv")
brand_dist = brand_bins.value_counts().sort_index()
brand_dist.to_csv("tables/brand_rating_distribution.csv")
print("Brand rating distribution saved")

brand_dist.plot(kind="bar", color="skyblue", edgecolor="black")
plt.title("Brand Rating Distribution")
plt.xlabel("Average Rating Range")
plt.ylabel("Number of Brands")
plt.savefig("figs/brand_rating_distribution.png", dpi=300)
plt.close()
print("Brand distribution chart saved")

#2.Category
#Categorize the categories according to the categories bands and plot a graph showing the distribution
category_ratings = master_unique.groupby("category")["average_product_rating"].mean().sort_values(ascending=False)
category_ratings.to_csv("tables/all_categories.csv")
print("Saved: tables/all_categories.csv")
bins = [0, 1, 2, 3, 4, 5]
labels = ["0–1", "1.1–2", "2.1–3", "3.1–4", "4.1–5"]
cat_bins = pd.cut(category_ratings, bins=bins, labels=labels, include_lowest=True)
cat_counts = cat_bins.value_counts().sort_index()
cat_counts.to_csv("tables/category_counts_by_rating_range.csv")
print("Saved: tables/category_counts_by_rating_range.csv")

plt.figure(figsize=(8, 5))
cat_counts.plot(kind="bar", color="skyblue", edgecolor="black")
plt.title("Number of Categories by Average Rating Range")
plt.xlabel("Average Rating Range")
plt.ylabel("Number of Categories")
plt.savefig("figs/category_counts_by_rating_range.png", dpi=300)
plt.close()
print("Saved: figs/category_counts_by_rating_range.png")

#3.Products by average ratings and number of reviews
top_products = (master_unique.groupby("name_title").agg({"average_product_rating": "mean","total_number_reviews": "sum"})
    .sort_values(by=["average_product_rating", "total_number_reviews"], ascending=False))
top_products.to_csv("tables/all_products.csv")
print("Full brand, category, and product data saved to tables/ folder")
plt.figure(figsize=(8, 6))
plt.scatter(
    top_products["total_number_reviews"],
    top_products["average_product_rating"],
    alpha=0.6,
    color="skyblue",       
    edgecolors="black")
plt.title("Products: Average Rating vs Total Number of Reviews")
plt.xlabel("Total Number of Reviews (Popularity)")
plt.ylabel("Average Product Rating (Quality)")
plt.grid(axis="both", alpha=0.3)
plt.savefig("figs/products_rating_vs_reviews.png", dpi=300)
plt.close()
print("Saved: figs/products_rating_vs_reviews.png")

Saved: tables/all_brands.csv
Brand rating distribution saved
Brand distribution chart saved
Saved: tables/all_categories.csv
Saved: tables/category_counts_by_rating_range.csv
Saved: figs/category_counts_by_rating_range.png
Full brand, category, and product data saved to tables/ folder
Saved: figs/products_rating_vs_reviews.png


Most brands and categories have average ratings between 2.1–3 and 3.1–4.
Only a few brands, like Work King and Nostalgia, have perfect scores, while others such as Young Land and Hello Kitty rate lowest.
Popular categories include windows and comfort sneakers, while areas like health & wellness get lower reviews.
Most products have fewer than 200 reviews, and only a few are both highly rated and popular.
Some items sell well but get lower ratings, suggesting possible quality issues.

In [9]:
#Are discounts driving sales or satisfaction? i.e. whether the discounts impact ratings
import pandas as pd
import matplotlib.pyplot as plt
# Load data
master_unique = pd.read_csv("jcpenney_master_dataset_unique.csv")
# Create discount column
master_unique["Discount_%"] = round(
    ((master_unique["list_price"] - master_unique["sale_price"]) / master_unique["list_price"]) * 100, 2)
# Plot the scatter graph
plt.figure(figsize=(8, 5))
plt.scatter(
    master_unique["Discount_%"],
    master_unique["total_number_reviews"],
    alpha=0.5,
    color="skyblue",     
    edgecolors="black")
plt.title("Relationship Between Discount (%) and Number of Reviews (Proxy for Sales)")
plt.xlabel("Discount (%)")
plt.ylabel("Total Number of Reviews")
plt.savefig("figs/discount_vs_reviews.png", dpi=300)
plt.close()
print("Saved: figs/discount_vs_reviews.png")
# Find the correlation
corr_sales = master_unique[["Discount_%", "total_number_reviews"]].corr().iloc[0, 1]
print(f"Correlation between discount and sales (reviews): {corr_sales:.2f}")

Saved: figs/discount_vs_reviews.png
Correlation between discount and sales (reviews): -0.03


Observations from the above analysis

The correlation between dicounts and sales (reviews) is -0.03 which is approximately 0. This suggests that bigger discounts do not have any impact on sales. JCPenney might need to focus on product quality or marketing instead of just lowering prices.

In [17]:
#Which age-group or states have a higher satisfaction?
bands = [0, 2, 3, 4, 5]
labels = ["0–2 (Low)", "2.1–3 (Avg)", "3.1–4 (Good)", "4.1–5 (High)"]
colors_blue = ["#87CEEB", "#5DADE2", "#3498DB", "#2E86C1"]

# a) States → satisfaction bands
state_satisfaction = (master_unique.groupby("State")["average_product_rating"].mean().round(2))
state_band = pd.cut(state_satisfaction, bins=bands, labels=labels, include_lowest=True)
state_band_counts = state_band.value_counts().sort_index()
state_band_counts = state_band_counts[state_band_counts > 0]
state_band_counts.to_csv("tables/state_satisfaction_bands.csv")
print("Saved: tables/state_satisfaction_bands.csv")
# Pie chart
plt.figure(figsize=(5,5))
plt.pie(
    state_band_counts.values,
    labels=state_band_counts.index,
    autopct="%1.0f%%",
    startangle=90,
    colors=colors_blue,
    pctdistance=0.85,      
    labeldistance=1.3,
    wedgeprops={"edgecolor": "white"})
plt.title("States by Satisfaction Band")
plt.tight_layout()
plt.savefig("figs/state_satisfaction_bands_pie.png", dpi=300)
plt.close()
print("Saved: figs/state_satisfaction_bands_pie.png")

# B) Age groups → satisfaction bands
# Ensure Age group exists
bands = [0, 2, 3, 4, 5]
labels = ["0–2 (Low)", "2.1–3 (Avg)", "3.1–4 (Good)", "4.1–5 (High)"]
age_group_rating = (master_unique.groupby("Age group", observed=False)["average_product_rating"].mean().round(2))
age_band = pd.cut(age_group_rating, bins=bands, labels=labels, include_lowest=True)
age_band_counts = age_band.value_counts().sort_index()
age_band_counts = age_band_counts[age_band_counts > 0]
age_band_counts.to_csv("tables/age_group_satisfaction_bands.csv")
print("Saved: tables/age_group_satisfaction_bands.csv")

# Donut chart
colors_for_pie = colors_blue[:len(age_band_counts)]
fig, ax = plt.subplots(figsize=(5,5))
ax.pie(
    age_band_counts.values,
    labels=age_band_counts.index,
    autopct="%1.0f%%",
    startangle=90,
    colors=colors_blue,
    pctdistance=0.8,
    labeldistance=1.1,
    wedgeprops={"width": 0.38, "edgecolor": "white"})
ax.set_title("Age Groups by Satisfaction Band")
plt.tight_layout()
plt.savefig("figs/age_group_satisfaction_bands_donut.png", dpi=300)
plt.close()
print("Saved: figs/age_group_satisfaction_bands_donut.png")

Saved: tables/state_satisfaction_bands.csv
Saved: figs/state_satisfaction_bands_pie.png
Saved: tables/age_group_satisfaction_bands.csv
Saved: figs/age_group_satisfaction_bands_donut.png


72% of the states fall in the “Average” satisfaction band (2.1–3), with 28% in the “Good” band (3.1–4), showing moderate satisfaction overall.

All age groups lie in the “Average” band, indicating that customer satisfaction is fairly consistent across ages and locations.

In [18]:
#What are customers saying?
#Convert to lowercase and remove missing values
reviews_text = master_unique["Review"].dropna().str.lower()
#Split words into text and count which words appear the most
#Import counter to count how many times a word appears, import re to remove the punctuations, etc, import stopwords to remove very common English words
from collections import Counter
import re
from nltk.corpus import stopwords
import nltk
import warnings
warnings.filterwarnings("ignore", category=UserWarning, module='nltk')
#Download stopwords
nltk.download("stopwords", quiet=True)
#Define stopwords set
stop_words = set(stopwords.words("english"))
#Combine all reviews into one big string
all_text = " ".join(master_unique["Review"].dropna().str.lower())
#Remove punctuation and numbers
all_text = re.sub(r"[^a-z\s]", "", all_text)
#Split into words and filter out stopwords
filtered_words = [word for word in all_text.split() if word not in stop_words]
#Count most common words
word_counts = Counter(filtered_words)

#Visualize it
from wordcloud import WordCloud
import matplotlib.pyplot as plt

wordcloud = WordCloud(width=800, height=400, background_color="white").generate(" ".join(filtered_words))

plt.figure(figsize=(10,5))
plt.imshow(wordcloud, interpolation="bilinear")
plt.axis("off")
plt.title("What Customers Are Saying", fontsize=14)
plt.tight_layout()
plt.savefig("figs/customer_reviews_wordcloud.png", dpi=300)
plt.close()

Observations from the above analysis:

The most common words in customer reviews include love, color, fit, great, and shoe. This shows that customers mostly talk about product appearance, comfort, and satisfaction. The frequent use of positive words like love and great suggests that most reviews are generally positive, though customers also focus heavily on fit and style details.

In [19]:
#How loyal or active are users?
# Count number of reviews per user
user_activity = master_unique["Username"].value_counts()
#Plot the graph
import matplotlib.pyplot as plt
plt.figure(figsize=(8,5))
plt.hist(user_activity, bins=20, color="skyblue", edgecolor="black")
plt.title("Distribution of User Activity (Number of Reviews per User)")
plt.xlabel("Number of Reviews Written")
plt.ylabel("Number of Users")
plt.grid(axis="y", alpha=0.3)
plt.savefig("figs/user_activity_distribution.png", dpi=300)
plt.close()
total_users = len(user_activity)
avg_reviews = user_activity.mean()
max_reviews = user_activity.max()

print(f"Total unique users: {total_users}")
print(f"Average reviews per user: {avg_reviews:.2f}")
print(f"Most active user wrote {max_reviews} reviews")

Total unique users: 4993
Average reviews per user: 7.83
Most active user wrote 68 reviews


Observations from the analysis above:

Total unique users are 4993 and the average reviews per user is 7.83. And most active user wrote 68 reviews. Most users wrote fewer than 10 reviews, with only a small number posting more. This shows that while many customers engage occasionally, only a few are highly active reviewers. The user with 68 reviews is an outlier in this case.